### PR TITLE
[Merged by Bors] - Add headless mode

### LINKED ISSUE #3155: Add headless renderer
## What problem does this solve or what need does it fill?

When creating multiplayer, often the logic of the game is executed on the headless server. But registration of renderer types such as `Asset<Meshes>` is now tied to the renderer. Therefore, to use exactly the same logic on the server, I have to add compile-time checks:

```rust
    pub fn spawn(
        commands: &mut Commands,
        transform: Transform,
        #[cfg(not(feature = "headless"))] meshes: &mut Assets<Mesh>,
        #[cfg(not(feature = "headless"))] materials: &mut Assets<StandardMaterial>,
    ) -> Self {
    // ...
    }
```

This looks ugly and very inconvenient to use.

## What solution would you like?

It would be great to have a headless renderer.

### PR DESCRIPTION
**Pull request by Shatur**
# Objective

In this PR I added the ability to opt-out graphical backends. Closes #3155.

## Solution

I turned backends into `Option` ~~and removed panicking sub app API to force users handle the error (was suggested by `@cart`)~~.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@cart do you want this in the 0.6 milestone? I could go either way on it I think. Super useful and rendering-linked, but reasonably complex.

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -853,14 +853,6 @@ impl App {
         self
     }
 
-    /// Retrieves a "sub app" stored inside this [App]. This will panic if the sub app does not exist.
-    pub fn sub_app_mut(&mut self, label: impl AppLabel) -> &mut App {
-        match self.get_sub_app_mut(label) {
-            Ok(app) => app,
-            Err(label) => panic!("Sub-App with label '{:?}' does not exist", label),
-        }
-    }
```
Comment (bjorn3) [inline crates/bevy_app/src/app.rs:862]: Why did you remove these functions?

Comment (Shatur) [reply crates/bevy_app/src/app.rs:862]: To force users handle the error. Was suggested by cart in https://github.com/bevyengine/bevy/issues/3155#issuecomment-987278686

File: examples/shader/shader_defs.rs [hunk 2]
```diff
@@ -23,7 +23,8 @@ pub struct IsRedPlugin;
 impl Plugin for IsRedPlugin {
     fn build(&self, app: &mut App) {
         app.add_plugin(ExtractComponentPlugin::<IsRed>::default());
-        app.sub_app_mut(RenderApp)
+        app.get_sub_app_mut(RenderApp)
+            .unwrap()
```
Comment (cart) [inline examples/shader/shader_defs.rs:27]: adding an unwrap here sort of defeats the point of using `get_sub_app_mut` over `sub_app_mut`

Comment (Shatur) [reply examples/shader/shader_defs.rs:27]: What should I do then?

Comment (Shatur) [reply examples/shader/shader_defs.rs:27]: @cart just to be clear: I removed `sub_app_mut` as you suggested to force users handle the error.

Comment (Shatur) [reply examples/shader/shader_defs.rs]: But panicking API back.

File: crates/bevy_app/src/app.rs (hunk 1 above)
Comment (Shatur) [reply crates/bevy_app/src/app.rs]: I put panicking API back.

**Comment by cart** [comment]
This breaks examples like `3d_scene` because it tries registering ImagePlugin before the RenderApp is inserted

**Comment by Shatur** [comment]
> This breaks examples like 3d_scene because it tries registering ImagePlugin before the RenderApp is inserted

Yeah, thinking about elegant solution. The function is huge.
